### PR TITLE
Added fuzzy_rename and fuzzy_duplicate, same as #31

### PR DESCRIPTION
**Pull request by mg979**

### CODE REVIEW

**Comment by facelessuser** [comment]
Thanks! I'll take a look at this sometime soon.

**Comment by mg979** [comment]
Ok no problem, when you have time.

File: fuzzy_file_nav.py [hunk 1]
```diff
@@ -613,7 +613,7 @@ def run(self):
                     os.remove(full_name)
             except Exception:
                 errors = True
-                error("Error deleting %d!" % full_name)
+                error("Error deleting %s!" % full_name)
```
Comment (facelessuser) [inline fuzzy_file_nav.py]: Ugh.  I copied and pasted that broken line everywhere 🙁 .

**Comment by facelessuser** [comment]
Rename should be in place, not "rename to" which is essentially "move file".  I'm not sure my thoughts on "duplicate to" yet, but I am probably more okay with that than I am with rename moving a file.

File: fuzzy_file_nav.py [hunk 2]
```diff
@@ -754,6 +754,133 @@ def run(self):
...
+        """Finalize duplicate."""
+
+        if name == self.basename:      # same name, do nothing
+            self.on_finish()
+            return
+
+        new = self.dir + os.sep + name
+        try:
+            if self.is_dir:
+                shutil.copytree(self.full_name, new, symlinks=True)
```
Comment (facelessuser) [inline fuzzy_file_nav.py]: Why symlinks True?  We don't do this anywhere else.

**Comment by facelessuser** [comment]
Now I remember why I never added this.  It's because it isn't needed.

Steps to rename:

1. Cut shortcut
2. Type new name in fuzzy panel
3. Paste shortcut

Steps to duplicate:

1. Copy shortcut
2. Type new name in fuzzy panel
3. Paste shortcut

There is no need to add a duplicate or rename method.

**Comment by facelessuser** [comment]
Rename on paste is also covered in the docs, so it was mentioned (I thought maybe it wasn't). If the docs aren't clear on this subject, let me know and I'll update the docs: https://facelessuser.github.io/FuzzyFileNav/usage/#fuzzy-file-paste.

With that said, I've fixed the bugs you found in this pull though.  So error messages are no longer using `%d`, and I've fixed the copy paste issue where notify was referencing `ExportHtml`.

**Comment by mg979** [comment]
Ok it makes sense, even if it's not the most user-friendly behaviour for me, but if you prefer. I see a small problem though:

> Steps to rename:
> 
> 1. Cut shortcut
> 2. Type new name in fuzzy panel
> 3. Paste shortcut

The way I did it, if the current view is the renamed file, it is set to scratch, closed and opened again with the new name. If the renamed file is not the current view but among the open views, the same, with focus that is given back to the current view after that.

With the current system, if you operate on open files, they keep their buffer and become dirty since there is no file anymore. So there should be at least a similar check with cut/paste operations.

Having a dedicated command, in my opinion makes it easier to use, though. Same for duplicate, even if duplicate doesn't have this problem.

**Comment by mg979** [comment]
Unrelated to this, I had a problem with the file `usage.md` in the dev/2.0 branch, Sublime says that the `worker xxx` had trouble processing the file and it's killing the process(but it doesn't), CPU usage goes high with multiple ST open processes that I had to kill manually, and if I open the file ST crashes. So there's some problem with that file in my Sublime Text(3143, Win10).

**Comment by facelessuser** [comment]
> The way I did it, if the current view is the renamed file, it is set to scratch, closed and opened again with the new name. If the renamed file is not the current view but among the open views, the same, with focus that is given back to the current view after that.

Yeah, but this adds two commands that can only be used on the current view. Do we then add a delete that operates only on the current view etc?  Personally, I just use bind the `Fuzzy Nav Here` to <key>ctrl</key>+<key>o</key> (shortcuts will be for a Windows example) and use that to open command in the current view, press <key>ctrl</key>+<key>.</key> to quickly enter the current file name and then copy and paste or cut and paste. Once the file is in the clipboard, you can navigate else where and paste there or paste multiple times.  And the 2.0 version remembers that file in clipboard across sessions.

The commands are designed to be general.  You navigate a directory and use these commands to work on any file.  I guess I'm not too keen to add a bunch of shortcuts that implement duplicate functionality, but **only** for the current view.  There is no need for a FuzzyNav panel for such commands. Here you are just creating a rename command and binding it to a shortcut.  But that functionality doesn't make sense in **this** package.

Maybe you just want something like https://packagecontrol.io/packages/File%20Rename? I've never used the package, but something like that *should* give you what you want.

**Comment by facelessuser** [comment]
> Unrelated to this, I had a problem with the file usage.md in the dev/2.0 branch, Sublime says that the worker xxx had trouble processing the file and it's killing the process(but it doesn't), CPU usage goes high with multiple ST open processes that I had to kill manually, and if I open the file ST crashes. So there's some problem with that file in my Sublime Text(3143, Win10).

It's just a markdown file.  Maybe you have a plugin trying to do something with it and is hanging because it has a bug.  It's just text.

**Comment by facelessuser** [comment]
Also, if Sublime is choking on the Markdown file with no plugins, then You may need to file a Sublime issue as it should be able to process any text file.

**Comment by facelessuser** [comment]
With the current file think **maybe** we could add logic to see if the file we are pasting is the current view and so a little extra cleanup to refocus the view. I'd probably be more open to that.

**Comment by mg979** [comment]
> Yeah, but this adds two commands that can only be used on the current view. Do we then add a delete that operates only on the current view etc?

Not really, the commands operate on any file. That behaviour (close if open and reopen) only kicks in if the files are among open views, and I think it's rather common. That kind of check doesn't need extra command though. Just a check among open views when a `move` command (copy/cut etc) is issued.

> Maybe you just want something like https://packagecontrol.io/packages/File%20Rename? I've never used the package, but something like that should give you what you want.

I don't want to use an extra package, again the point is not renaming open files only, just any file. I'll use the steps you suggested, and I guess I'll make a couple of extra commands for my personal use and keep them in another file :).

**Comment by facelessuser** [comment]
That makes sense. I use TabsExtra which adds a duplicate and rename in the context tab if I really want to rename quick.

I would also make sure to not blindly close a file with unsaved changes. I'd want to be notified if I was careless with option to cancel. Maybe you did that, I don't know.

**Comment by mg979** [comment]
> That makes sense. I use TabsExtra which adds a duplicate and rename in the context tab if I really want to rename quick.

I don't use much tabs, I prefer having a single way to take care of these operations.

> I would also make sure to not blindly close a file with unsaved changes. I'd want to be notified if I was careless with option to cancel. Maybe you did that, I don't know.

No I wasn't doing it but I agree it should be done.

As for the rest, I think they are two different things, the extra commands and the open files check. The second could be done also for copy/cut commands if you don't want extra commands.

Something on the Iine of what I was using this for rename

```python
	# close and open renamed file if among open files
	current_view = self.window.active_view()
	current_file = current_view.file_name()
	for v in self.window.views():
	    if v.file_name() == self.full_name:
	        self.window.focus_view(v)
	        v.set_scratch(True)
	        self.window.run_command('close_file')
	        self.window.open_file(new)
	        if current_file != self.full_name:
	            sublime.set_timeout(lambda: self.window.focus_view(current_view), 100)
        break 
```

**Comment by facelessuser** [comment]
> I don't use much tabs, I prefer having a single way to take care of these operations.

Understood

I had thought about adding rename and duplicate before.  I just had a hard time justifying it.  You don't really gain much.  With both implementations (what is currently in place and what you proposed) requires you to target a file, enter a new name, and use a key stroke to finalize it (enter key in your case).

The one difference in yours essentially attempts a homegrown retarget on the view, and my approach just stores the file in the clipboard so it can be pasted in different places.  Though there is no reason the retarget couldn't be added to cut paste that relates to an open view.

If there is one thing I think I actually desire, it is maybe storing multiple files in the clipboard.  That is why I have the clipboard as an array, I just haven't bothered to implement pushing multiple files onto the clipboard stack yet. Obviously things like "paste as" won't work when the clipboard has multiple files in the stack.

On a side note, sublime also has a retarget API command to essentially reopen the window with its new target location:

```
            v = self.window.find_open_file(old)
            if v:
                v.retarget(new)
```

**Comment by mg979** [comment]
> If there is one thing I think I actually desire, it is maybe storing multiple files in the clipboard. That is why I have the clipboard as an array, I just haven't bothered to implement pushing multiple files onto the clipboard stack yet.

To achieve what, exactly? To copy a sequence of files somewhere else without repeated copy/cut/paste? You could have an extra command for that, without changing default behaviour. Like `ctrl+m`, `fuzzy_mark`, it wouldn't need tab completion, just get the index form the panel and keep the panel open, it adds to a list and then you do with that list what you need to do.

I didn't know about the retarget function, it could work too.

**Comment by mg979** [comment]
I think a different command would be better or it could get confusing. Now you have to press tab, copy, then paste. Imagine you have a command `fuzzy_mark_for_copy`, bound to `ctrl+shift+c` for example. You browse the quick panel and hit `ctrl+shift+c`, without pressing tab first. The path of the file in the index is copied to your list. Then another command `fuzzy_paste_stack`, bound to `ctrl+shift+v` for example, that lets you paste the stack in that place.

By the way I made a new branch based on current dev/2.0, I moved extra commands to a different file, so that the main file isn't touched. It includes:

- rename and duplicate
- send to trash (bound to ctrl+alt+d for now), it needed an additional module so it's there too
- a Default.sublime-keymap for the extra commands

https://github.com/mg979/FuzzyFileNav/tree/dev_mg979

**Comment by facelessuser** [comment]
> send to trash (bound to ctrl+alt+d for now), it needed an additional module so it's there too

Sublime actually has a send2trash included in default.  I'd probably just leverage that.  And maybe make it default, I've had that on my list to look into as well.

> I think a different command would be better or it could get confusing. Now you have to press tab, copy, then paste. Imagine you have a command fuzzy_mark_for_copy, bound to ctrl+shift+c for example. You browse the quick panel and hit ctrl+shift+c, without pressing tab first. The path of the file in the index is copied to your list. Then another command fuzzy_paste_stack, bound to ctrl+shift+v for example, that lets you paste the stack in that place.

As mentioned, adding more files to the stack would be a different command.  Copy would clear the clipboard and add the file.  Copy append would add a file to the stack.  And paste just pastes anything in the clipboard stack.  If you have multiple files in the stack and you try to paste with a rename, it would alert you that it can't do that.  I don't think that wold be too confusing.  I just need to check how I currently deal with the stack and make adjustments.

**Comment by facelessuser** [comment]
Permanent delete could maybe get moved to a separate command.

**Comment by facelessuser** [comment]
Probably could show status bar status of how many files in the copy stack as well....

**Comment by mg979** [comment]
I can imagine it like this:

- copy wouldn't need tab completion, it would get the files list from `FuzzyFileNavCommand.files`, and the index from `on_higlighted` function of the quickpanel.

- after a copy (or at the second copy command), a bottom panel shows up, where you can track the current files/dirs stack. The copied paths are added to the list.

- when you want to paste: if there is only one item in the stack, it's pasted without prompt, and the list is cleared. If there are more items, a quick panel is shown, with:

  - paste last (keep the stack)
  - paste selected (show another panel where you can choose which item to paste, keep the stack)
  - paste all (keep the stack)
  - paste all (clear the stack)
  - clear the stack and paste nothing

**Comment by facelessuser** [comment]
> copy wouldn't need tab completion, it would get the files list from FuzzyFileNavCommand.files, and the index from on_higlighted function of the quickpanel.

The tab completion is pretty much an established convention.  You can select your tab completion style, but it is likely to not go away.

> after a copy (or at the second copy command), a bottom panel shows up, where you can track the current files/dirs stack. The copied paths are added to the list.

Like an output panel?  I guess that could be an option.  I likely would only want to view something like that on a rare occasion, but it couldn't hurt if people want to see it.

> when you want to paste: if there is only one item in the stack, it's pasted without prompt, and the list is cleared. If there are more items, a quick panel is shown, with:
>
> - paste last (keep the stack)
> - paste selected (show another panel where you can choose which item to paste, keep the stack)
> - paste all (keep the stack)
> - paste all (clear the stack)
> - clear the stack and paste nothing

- I don't see a reason to clear the stack if pasting one file. I may want to paste it again.
- When you copy again, the stack will be cleared.  But just pasting after a copy doesn't make sense. After cut and paste, the stack should get cleared (I'm probably not doing this right now), but I should ensure that is occurring. Also, a clear should occur on transition from copy to cut and cut to copy.  You shouldn't be able to append cut files to copy and copy files to cut.
- I should add a manual clear to the stack though. Especially if you've cut and decided against pasting.  That way you don't accidentally paste and remove the original.
- The idea of selectively pasting items from the stack is an interesting idea though.  I will give that some thought.

**Comment by mg979** [comment]
Ok on everything, but..

> When you copy again, the stack will be cleared.

And if you wanted to build a stack, paste it, and then wanted to add a file to the stack and keep it? After you build a stack, automatic clearing on subsequent copy wouldn't be the expected behaviour for me.
In my opinion the stack clearing should be explicit, either in a paste panel option, or a prompt after you pasted something, if the stack is not a single item. Like, paste a stack and you're asked if you want to keep the stack or not.

> But just pasting after a copy doesn't make sense.
Not understanding.

> Also, a clear should occur on transition from copy to cut and cut to copy. You shouldn't be able to append cut files to copy and copy files to cut.

You could make the list as a list of tuples `(path, bool)`, so that you can discern the kind of command when pasting. I don't think it would be that common to mix commands, but you could do it. In a quick/output panel the kind of command would also be listed. And personally I would avoid automatic clearing, but I don't know.

**Comment by facelessuser** [comment]
> And if you wanted to build a stack, paste it, and then wanted to add a file to the stack and keep it? After you build a stack, automatic clearing on subsequent copy wouldn't be the expected behaviour for me.

This won't be a problem. Yes you will be able to force a clear, but there will be two different copy commands: "Copy" and "Copy Append".  "Copy" starts a stack while "Copy Append" appends to the stack.  The same would go for "Cut" except that a "Cut" clears the stack on paste. So with a "Cut", you could "Cut Append" up until paste, then the stack will be empty and additional "Cut Appends" would be to a new empty stack.

**Comment by facelessuser** [comment]
> You could make the list as a list of tuples (path, bool), so that you can discern the kind of command when pasting. I don't think it would be that common to mix commands, but you could do it. In a quick/output panel the kind of command would also be listed. And personally I would avoid automatic clearing, but I don't know.

FuzzyClipboardCommand already tracks clips and action, so this should be easy enough to tap into.

**Comment by mg979** [comment]
Understood.

> there will be two different copy commands: "Copy" and "Copy Append". "Copy" starts a stack while "Copy Append" appends to the stack. The same would go for "Cut" except that a "Cut" clears the stack on paste

The choice is yours, but I don't think that there's the need to clear the stack on Cut. For example:

- copy a single item, and add to the stack (path, "copy"). you paste it, it stays in the stack.
- copy something else, it's added to the stack. if you paste it, you're asked if you want to keep the stack or not.
- you keep it. cut something, and it adds (path, "cut") to the stack.
- you call the output panel just to check, and it says
```
FuzzyFileNav: current stack:
path1            copy
path2            copy
path3            cut
```
- you paste a selected item, you choose the cut item. you're asked if you want to keep the stack, say yes, the cut item is removed, the copied items stay.
- or you choose to paste all, you the choose to keep the stack, say yes, the cut item is removed, but the others stay.

That is, on paste command, cut items would be removed, and copied items not.

**Comment by facelessuser** [comment]
I think mixing cuts and copies in the stack is a weird messy workflow.  In any other application you can select multiple files for copy **or** multiple files for cut.  Yes you can look at the stack with a special command to see what is what, but I don't want to design this where people **have to** leave the stack fully visible to remember what is what. To be honest most people would just cut and paste or copy and paste and not try and mix it as that is just how people's brains usually work.  I have a task so I will complete it. No sense adding complexity just in case someone likes to plan for a cut when they haven't finished their copy.

**Comment by mg979** [comment]
Ok then, it was just an idea, sorry.

**Comment by facelessuser** [comment]
No need to apologize.  Did I come off too critical?  I'm just giving my impression on it.  Ideally I like to keep the code as simple as I can and only introduce complexity where it makes sense.

**Comment by mg979** [comment]
No problem, I realize it's better to keep it straightforward. An output panel would be cool, though, and the ability to select a single item from the list as well.

**Comment by facelessuser** [comment]
Keep in mind that the selective paste is probably a bit further out than multiple paste. My view of things can often shift the more I use things.

I'm not ruling out the possibility of adding mixed cut and copy in the future, only that my general first impression is avoid it. Things can evolve though. I often argue against ideas until I'm convinced otherwise. Just because I initially dislike something doesn't mean I've closed the door to it forever.

**Comment by facelessuser** [comment]
Copying multiple files is on the 2.0 branch. I even added selective paste which I wasn't planning on.  Couple of general fixes as well.  Anyways, there may be some bugs, but the initial work is there.

**Comment by mg979** [comment]
Ok thanks, I'll update the branch, I was working on a thing and I have to fix one issue, when it's ready I'll show you, maybe you'll like it.

**Comment by facelessuser** [comment]
Still don't have manual clipboard clear, but maybe later today. It will be a normal command with probably no key binding on the fuzzy panel, but maybe it will. Not sure what the binding would be if I do add it, but not sure if I need it either.

**Comment by mg979** [comment]
This is a bit of what I've been doing:

Imgur

Basically, all commands can also be accessed by menu, if you call it instead of tab completion. There are two kinds of menus:

- `ctrl+k, ctrl+tab` opens a menu with regular commands, works on both files and dirs
- `ctrl+k, ctrl+f` only works on dirs and has other commands:
  - open all files in directory
  - close all files in directory if open and not dirty
  - add/remove folders to project
  - add/remove folder to fuzzy bookmarks

Also integrated `send to trash`/`delete `in the same command.

If you like the idea, I can make another of these for appended clips, that shows also the current stack, for example if it's a copy stack, on `ctrl+a` a quick panel:

```
  Copy (append)
  
  Current stack:
  ------------------------------
  item1
  item2
  ...
```

This would also make the output panel superfluous.

**Comment by facelessuser** [comment]
It may be a good idea to add something like this, so in general I think I like the concept.

Here are my more nit picky thoughts as well to give you an idea as how I would prefer it to be approached:

- Instead of having two separate shortcuts that open a menu for folders vs files, I'd prefer to make it something that is context aware with one simple shortcut (not a big fan of using double shortcut sequences if I can avoid it).  When applied to a file, it would show file related options, and folders would show folder related options.
- As mentioned before, I think I am cool with leveraging Sublime's send2trash for default deletes, and providing a command for permanent delete if desired.  So I think this is fine.
- I think removing a folder from a project is a fine addition as adding already exists.  The commands  would be context aware and should only show if the folder is in the project or not in the project respectively.
- I think adding/removing a folder from bookmarks is also fine. This would probably be context aware like adding/removing folders to projects would be.
- I'd probably pass on the open all files in the directory, but I'm not sure yet.
- Definitely pass on the close all files in the folder as I just don't manage tabs that way and have never felt the need for it.
- > assign custom project folder name for sidebar ("name" in projects data folders)

    I'm not sure what this is.

- I wonder if we even need a separate command that shows the clipboard append history as when you hit ctrl+shift+v, you are presented with the clipboard stack and can select options to paste or hit esc to cancel if you just wanted to view.  So it kind of serves as a viewer and a paster.  I guess it's possible to augment the command to take an action that shows it in a view only state, but I'm not sure if that is needed.

- I'd probably ditch the header though on the panel.

    ```
    doc
    -------------
    ```

    I'm not a big fan of having options in the panel that are just for aesthetic purposes and do nothing practical.

- Probably won't change shortcuts to match what is shown, but I imagine this is just your personal setup.

Sublime 3154+ adds an interesting new API for command panel command inputs that I may play with.  This may affect how new features are added.

**Comment by mg979** [comment]
> Instead of having two separate shortcuts that open a menu for folders vs files, I'd prefer to make it something that is context aware with one simple shortcut (not a big fan of using double shortcut sequences if I can avoid it). When applied to a file, it would show file related options, and folders would show folder related options.

I was doing it like this at first, but the problem is that the same actions you can perform on files, you can perform them on dirs too, but dirs have exclusive actions, and if I mix them all the panel is a bit cluttered when handling dirs, but I can change it back.

> I think removing a folder from a project is a fine addition as adding already exists. The commands would be context aware and should only show if the folder is in the project or not in the project respectively.
> I think adding/removing a folder from bookmarks is also fine. This would probably be context aware like adding/removing folders to projects would be.

Yes they are already context aware (showing `remove` if already in project/bookmarked, `add` otherwise).

> I'd probably pass on the open all files in the directory, but I'm not sure yet.
> Definitely pass on the close all files in the folder as I just don't manage tabs that way and have never felt the need for it.

Those were actually the n.1 reason why I started all this, because there is no way to open all files of a project folder in ST, not from the sidebar or anywhere. Same for close, I've been in the need of opening a dir with many files, doing a quick regex replace on all files, and then close them all. I had open them one by one, and close them one by one. I think it's an useful addition, but it's up to you. I could add a filter to manage project exceptions, so that `open` command behaves differently from simply dragging files from explorer.

> > assign custom project folder name for sidebar ("name" in projects data folders)

> I'm not sure what this is.

I added it because it was a possible option, but I don't really care about it. It would be the same as this:
https://packagecontrol.io/packages/FolderAlias%20Rename%20Tool

> I'd probably ditch the header though on the panel.

The header is there because calling this menu bypasses tab completion, so it tells you what file/dir you're working with, since it can't be in the input field. It can be removed anyway, it's not that necessary.
If you say it shouldn't bypass tab completion, I have to say that: many commands are a replica of what you can do with tab completion(except dir-specific commands), if you enforce tab completion for this menu it would mean one more step to achieve the same(when the menu is called, it passes the index to FuzzyPanelText as if tab completion had taken place).

For normal commands, I see it as an alternative way for people that prefer to see available commands. For dir commands, it's a way to avoid additional keybindings. But its usage shouldn't be more unpractical than necessary, I think.

> Probably won't change shortcuts to match what is shown, but I imagine this is just your personal setup.

Yes I didn't find yet a setup I'm happy with, but it's not important.

> Sublime 3154+ adds an interesting new API for command panel command inputs that I may play with. This may affect how new features are added.

I'm not registered yet, but I would avoid making options for registered users only, unless there's no other way of doing them. Are these features documented somewhere?

Also I'm curious where you found that send2trash can be handled by Sublime, and that retarget function, they aren't in the official API. Is there a place where I can look for undocumented features/functions? Thanks

**Comment by mg979** [comment]
In the last commit I also added a menu for the stack, clear/paste options mostly, also with `clear selected`

Imgur

**Pull request closed without merging**